Refactor csv export

chrwm opened the pull request:
Closes #400

Commit:
Remove StatistikFlag from orm.py #400

Commit:
Delete parameter statistik_flag #400

The feature has been depreciated from Markstammdatenregister

Commit:
Move csv export related functions to helpers.py #400

* rename self.--- objects
* move constants to constants.py

Commit:
Move constants to constants.py #400

Commit:
Move csv export to helpers.py #400

Commit:
Adjust mastr.py to new csv export #400

Commit:
Update script #400

* was not tested

Commit:
Update imports #400

* rename self.--- objects
* move constants to constants.py

Commit:
Refactor orm table and subtables selection #400

Commit:
Add parameter validation for csv_export #400

Commit:
Rename metadata creation function #400

Commit:
Refactor table join #400

Can not be improved as e.g. `orm.BasicUnit.EinheitMastrNummer == orm_tables["unit_data"].EinheitMastrNummer` EinheitenMastrNummer has to be called with .

Commit:
Merge additional table and tech table export functions #400

* all csv tables are exported via one export function
* all db queries are created in one function `create_db_query`. The structure of the function should enable easy tables joins in the future for csv export within on session.
* additional tables can be now also exported with a limit
* in constants.py ORM_MAP was extended. ORM_MAP maps the parameters that the user inserts to the tables classes in orm.py. This means additional tables for export can be easily added here, if desired.
* metadata saving was moved to its own function

Commit:
Specify index_label for csv export #400

Commit:
Add table naming for additional_tables #400

* generalize csv-export logging info

Commit:
Clean imports #400

Commit:
Update docstring of db_query_to_csv #400

Commit:
Update docstring of save_metadata #400

Commit:
Update docstring of create_db_query #400

Commit:
Update docstring of to_csv #400

Commit:
Add conftest.py to collect fixtures #400

Commit:
Refactor for easier unit testing #400

Commit:
Update test_db_query_to_csv #400

Commit:
Merge remote-tracking branch 'origin/develop' into refactor-400-csv-export

# Conflicts:
#	open_mastr/soap_api/mirror.py

Commit:
Update CHANGELOG.md #400

chrwm commented:
A short summary of the most important changes:
3 main functions exist:

1. csv-export is triggered via `MaStR.to_csv()`: defines and collects tables to export
2. `open_mastr.utils.helpers.create_db_query`: creates db-queries for TECHNOLOGIES and ADDITIONAL_TABLES
3. `open_mastr.utils.helpers.db_query_to_csv`: exports db-query to csv for TECHNOLOGIES and ADDITIONAL_TABLES

- functions relevant to the csv_export are now located in helpers.py


**Pytest**
Only open_mastr.utils.helpers.db_query_to_csv is tested for:
```
The test checks for 2 random tech and 2 random additional tables:
1. If csv's have been created and encoded in 'utf-8' and are not empty
2. For techs, if limited (limit=60) EinheitMastrNummer in basic_units are included in CSV file
3. For additional_tables, if csv is not empty
```
It is due to the Markstammdatenregister has unstable columns (columns are renamed, new columns are added, columns are deleted). This makes testing with static columns impractical in my opinion.
@FlorianK13 @deniztepe if you have ideas for better testing, please make suggestions.

Commit:
Skip test_db_query_to_csv if db is smaller than 1MB #400

chrwm commented:
Volkswagen was not necessary ;)

deniztepe commented on open_mastr/utils/config.py:
```diff
@@ -222,6 +222,11 @@ def _filenames_generator():
         )
 
+        # Add filenames for additional tables
+        filenames["raw"].update({"additional_table":
```
Since I already have a filenames.yml file. It is not updated and this led to an error during csv export. Manually removing the filenames and run the code works, but could we indicate or in the best case automate the update of config files? A user would get the same error after updating the package, right?

chrwm replied:
Yes, that's right. I thought about this but wanted to tackle it in another PR before the next release as this one is really full already.
A simple solution would be to remove the `if not` here https://github.com/OpenEnergyPlatform/open-MaStR/blob/46f51056636404e2c706ab9eb69bf164083f4d54/open_mastr/utils/config.py#L143-L151
However, this would lead to an update of the filenames.yml everytime the Mastr class is called and longer runtimes, as the function is called in `setup_project_home()` which is called in `__init__.py`.
Better solutions would check for e.g. the version or include refactoring. But both options should be handled in another issue IMO.

deniztepe replied:
I created a new issue #410. Your suggestion makes sense for me.

deniztepe commented on open_mastr/utils/helpers.py:
```diff
@@ -410,3 +429,329 @@ def data_to_include_tables(data: list, mapping: str = None) -> list:
+        mastr_technologies = [
+            unit_type_map_reversed[tech] for tech in data
+        ]
+        newest_date = (
```
When I run `mastr.to_csv(["balancing_area"] ` following error occurs
```
Traceback (most recent call last):
  File "C:\Users\deniz\Documents\open-MaStR\open_mastr\mastr_download_deniztest.py", line 59, in <module>
    mastr.to_csv(["balancing_area"])
  File "C:\Users\deniz\Documents\open-MaStR\open_mastr\mastr.py", line 368, in to_csv
    save_metadata(data=technologies_to_export, engine=self.engine)
  File "C:\Users\deniz\Documents\open-MaStR\open_mastr\utils\helpers.py", line 590, in save_metadata
    session.query(orm.BasicUnit.DatumLetzteAktualisierung)
TypeError: 'NoneType' object is not subscriptable
```
Because the basic_units table remains empty, when only additional tables are exported. Can we maybe add a try-except block wrapping this newest date and have a fallback date (maybe DatumDownload or dataversion date)?

chrwm replied:
This is rather because `technologies_to_export` is emtpy in case of additional_tables and the query fails. The backfill is performed before in https://github.com/OpenEnergyPlatform/open-MaStR/blob/46f51056636404e2c706ab9eb69bf164083f4d54/open_mastr/mastr.py#L341-L344

However, thanks for spotting. This will be tackled in https://github.com/OpenEnergyPlatform/open-MaStR/issues/386 and is already listed as action point there. Until then, I comment `save_metadata` out.

deniztepe replied:
Resolved with https://github.com/OpenEnergyPlatform/open-MaStR/pull/401/commits/dc883995de0ec41e678ddccc67f644bc5d21da68 at the moment.

deniztepe commented on tests/test_helpers.py:
```diff
@@ -1,17 +1,37 @@
+# Check if db is empty
+_db_exists = False
+_db_path = os.path.join(expanduser("~"), ".open-MaStR", "data", "sqlite", "open-mastr.db") # FIXME: use path in tmpdir when implemented
+if os.path.isdir(_db_path):
```
Either if  should be replaced with `os.path.isfile(_db_path)` or below in the for loop the entry must be checked and the path should be the sqlite directory. 

Currently,` _db_exists` is never going to be True, right?

chrwm replied:
True, I'd opt for changing the path, so that `os.scandir` can be used still.

deniztepe replied:
I renamed one parameter not to overwrite _db_path in 092383c088eed89b97189600895b328c64f2c285. If it seems correct for you, you can resolve conversation.

deniztepe reviewed (changes requested):
Overall it works well. I guess, the commented three points are small issues.

Commit:
Disable save_metadata #400

Currently metadata is only and can only be created for technologies. Will be enabled for additional_tables in #386

Commit:
Update db empty check #400

deniztepe committed:
Edit db empty check to scan folder #400

* _db_path was getting longer if the folder has more files
* black applied

chrwm commented:
@deniztepe you can approve, if I met your suggestions.

nesnoj commented:
Hey @chwrwm, I'd like to reference the package prior to the CSV refactoring. However, the latest version 0.12.1 does not include all recent fixes. 
It would be very nice if you could prepare an in-between-release (at 1ba8c7ce0b024ef7f65ba792e6a33021b89e5848) without the CSV stuff.
Is that possible for you? If not, I could take take of this..

nesnoj commented:
> Hey @chwrwm, I'd like to reference the package prior to the CSV refactoring. However, the latest version 0.12.1 does not include all recent fixes. It would be very nice if you could prepare an in-between-release without the CSV stuff. Is that possible for you? If not, I could take take of this..

Hmm, maybe it is not necessary. Did the CSV format change with the refactoring?
